### ISSUE TITLE
Rule 933100 FP on "<?xml" body

### ISSUE THREAD

**Issue by tomsommer** [SpiderLabs/owasp-modsecurity-crs#654]
Rule 933100 catches all `xmlrpc.php` requests which start with `<?xml`

**Comment by dune73** [comment]
Thanks for reporting. Could you provide a full request (request line + request body), please? This would kickstart our work.

**Comment by lifeforms** [comment]
I've noticed this if the client did not send the proper XML Content-Type. Do you have control over the client code making the request? In any case, a full log entry would be extremely useful to get to the bottom of this.

**Comment by tomsommer** [comment]
```
--8c52581f-B--
POST /xmlrpc.php?for=jetpack&token=xxxxxxxxxxxx&timestamp=1479979676&nonce=xxxxxx&body-hash=xxxxxxxxxxx%3D&signature=xxxxxxxxx%2FZtz14s%3D HTTP/1.1
Host: xxxxxxxxxxxxx
User-Agent: Jetpack by WordPress.com
Accept: */*
Accept-Encoding: deflate, gzip
Referer: xxxxx
Authorization: X_JETPACK xxxxxxxxxxxxxxxxxxx
Connection: close
Content-Length: 116
Content-Type: application/x-www-form-urlencoded

--8c52581f-C--
<?xml version="1.0"?>
<methodCall>
<methodName>jetpack.idcUrlValidation</methodName>
<params>
</params></methodCall>
--8c52581f-F--
```

Have to blank out some details, sorry

**Comment by lifeforms** [comment]
As I expected the Content-Type is incorrectly sent by the client (should be `application/xml`) and therefore ModSecurity's XML parser does not kick in, and we are left to deal with the raw content.

I've seen the same error in an in-house app. This error is probably widespread, so it's our problem :) Thanks for submitting it.

We should make our rule more robust so it doesn't get tripped up by this. We have to be careful for evasions though. I can propose a fix for 3.0.1.

**Comment by lifeforms** [comment]
By the way, for now you can temporarily disable this rule for xmlrpc.php with the following configuration (pick a unique id for yourself):

```
SecRule REQUEST_FILENAME "@streq /xmlrpc.php" \
	"id:12345,phase:1,t:none,nolog,pass,\
		ctl:ruleRemoveById=933100"
```

**Comment by dune73** [comment]
If JetPack from WP does this wrong, we could cover this in the WP rule exclusions. 

But you are right, that 933100 really has it coming: `@pm <? <?php ?> [php] [\php]`

**Comment by lifeforms** [comment]
Yep. Sucks. We can do something similar to rule 953120, probably with a space after xml, e.g. `<\?(?!xml )` or maybe `<\?(?!xml\s)` after reading the XML spec. Requiring the space prevents a well-known evasion like `<?xml_error_string(0);` which tries to look like an XML document header but is still valid PHP.

Also, `[\php]` looks wrong too, I guess this is supposed to be bbcode or something but then it should be `[/php]`.

**Comment by dune73** [comment]
Agreed.

**Comment by lifeforms** [comment]
Okay, this turns out to be tricker than I expected.

Naively converting the pm to regexp seems easy enough.

| Item | PM             | Regexp             |
|------|----------------|--------------------|
| 1    | `<?`           | `<\?`  |
| 2    | `<?php`        | `<\?php`           |
| 3    | `?>`           | `\?>`              |
| 4    | `[php] [\php]` | `\[(?:/|\\)?php\]`  |

(Github mangles item 4, don't mind it.)

We want to prevent item 1 from matching the `<?xml version="1.0"?>` XML declaration. That can easily be done with a negative lookahead: `<\?(?!xml\s)`. This prevents known evasions.

A bigger problem is item 3, the closing PHP tag. First, the value of checking for it is debatable because in most attacks detecting the opening tag is much more important. It will break out of text context into the PHP interpreter and that is very useful if you can get a vulnerable web app to write somewhere. Detecting the closing tag is less useful in my opinion. But, let's assume we do not want to remove functionality and try to detect `?>` as we have done.

Unfortunately, the XML declaration contains the `?>` string and gives a false positive for item 3 even if we have fixed item 1. So, if we want to match `?>`, we must make sure that this `?>` is *not* part of a possible *initial* `<?xml version="1.0"?>` XML declaration.

However, we must be very strict in rejecting the match *only* if the `<?xml...?>` appears at the *start* of the input string. Otherwise, we are open to injections such as:

```
echo '123'; // <?xml version="1.0"?>
foo(); #<?xml version="1.0"?>
```

These payloads end with a PHP comment so the PHP interpreter will ignore the xml-like tag, while picking up its own PHP closing tag.

But I found it very hard to do. Could be that my regexp skills are good but not the best. The best thing I could come up with in the time today was:

```
@rx (?:<\?(?!xml\s)|<\?php|^(?!<\?xml\s).*\?>|\[(?:/|\\\\)?php\])
```

This correctly gets rid of the XML FPs tested. It will pick up a lone `?>` at the start or in the middle of a string, but not if it's part of a starting `<?xml ....?>`. But this leaves open some, maybe VERY far fetched, injections like:

```
<?xml version="1.0"?>'; foo(); ?>
<?xml version='1.0'?>"; foo(); ?>
```

I could see this perhaps being exploitable in an extremely silly templating engine that is fed user input which is evalled in some `echo` context unescaped. Very far fetched, I have no actual examples for this, I could not see it happen in a reasonable application, but it's not feeling 100% safe. However maybe to a large degree. ;) In fact detecting the closing tag might not even be such an important part of the rule anyway; I think mostly it would help against smuggling PHP data into various file contents and then including them through some other RCE. Anyway... this XML FP really is common and a blight on the CRS. Comments are welcome. I might do some more thinking on this...

My tests:

```
num test sco actual-rules                 expected-rules         payload
--- ---- --- ---------------------------- ---------------------- -----------------------------
1   ok   10  941100,942370                !933100                <?xml version="1.0"?>
2   ok   10  941100,942370                !933100                <?xml version='1.0'?>...
3   ok   10  941100,942370                !933100                <?xml\n version="1.0"?>
4   ok   10  941100,942370                !933100                <?xml\n version="1.0" ?>
5   ok   10  941100,942370                !933100                <?xml\n version = "1.0" ?>
6   ok   10  941100,942370                !933100                <?xml version="1.0" encoding="UTF-8"?>
7   ok   13  941100,942370,942430         !933100                <?xml version="1.0" encoding="UTF-8" standalone="no" ?>
8   ok   10  941100,942370                !933100                <?xml version="1.0"?><foo></foo>
9   ok   5   933100                       933100                 <?
10  ok   5   933100                       933100                 <? x
11  ok   5   933100                       933100                 <? ?>
12  ok   5   933100                       933100                 <? blah();
13  ok   10  933100,942440                933100                 <? blah(); /*
14  ok   5   933100                       933100                 <? blah(); //
15  ok   10  933100,941100                933100                 <?xml() ?>
16  ok   5   933100                       933100                 <? xml(); ?>
17  ok   5   933100                       933100                 texttext<? blah();
18  ok   5   933100                       933100                 ?>
19  ok   10  932150,933100                933100                 echo 'hi'; ?>
20  ok   5   933100                       933100                 o?>
21  ok   8   933100,942110                933100                 "?>
22  ok   8   933100,942110                933100                 " ?>
23  ok   8   933100,942110                933100                 "no" ?>
24  ok   10  933100,942370                933100                 version="1.0"?>
25  ok   5   933100                       933100                 blah(); ?>
26  ok   5   933100                       933100                 foo(); // ?>
27  ok   5   933100                       933100                 <? blah(); ?>
28  ok   10  933100,941100                933100                 <?xml();
29  ok   15  933100,933151,941100         933100                 <?xml_error_string(0);
30  ok   10  933100,933151                933100                 <? xml_error_string(0);
31  ok   5   933100                       933100                 <?php
32  ok   5   933100                       933100                 <?PHP
33  ok   5   933100                       933100                 <?php blah();
34  ok   5   933100                       933100                 <?PHP blah();
35  ok   5   933100                       933100                 <?php blah(); ?>
36  ok   5   933100                       933100                 [php] blah();
37  ok   10  932105,933100                933100                 [php] blah(); [/php]
38  ok   10  932105,933100                933100                 blah(); [/php]
39  ok   5   933100                       933100                 blah(); [\php]
40  ok   8   933100,942110                933100                 '; foo(); ?>
41  ok   8   933100,942110                933100                 "; foo(); ?>
42  ok   20  933100,941100,942370,942440  933100                 /* <?xml version="1.0"?> */ foo(); ?>
43  ok   20  932150,933100,941100,942370  933100                 echo '123'; // <?xml version="1.0"?>
44  ok   15  933100,941100,942370         933100                 <?xml version="1.0"?><? foo(); //
45  ok   15  933100,941100,942370         933100                 <?xml version="1.0"?><? foo(); #
46  ok   15  933100,941100,942370         933100                 <?xml version="1.0"?><? foo(); ?>
47  ok   18  933100,941100,942370,942430  933100                 <?xml version="1.0"?><? foo(); ?><?xml version="1.0"?>
48  ok   20  933100,941100,942370,942440  933100                 x();#<?xml version="1.0"?>
49  ok   20  933100,933151,941100,942370  933100                 <?xml_error_string(0); // version="1.0"?>
50  ok   15  933100,942370,942440         933100                 <? xml(); # version="1.0"?>
51  FAIL 15  941100,942330,942370         933100                 <?xml version="1.0"?>'; foo(); ?>
52  FAIL 15  941100,942330,942370         933100                 <?xml version='1.0'?>"; foo(); ?>
53  ok   15  933100,941100,942370         933100                 garbage<?xml version="1.0"?>
54  ok   10  933100,942370                933100                 <?yml version="1.0"?>
55  ok   10  933100,942370                933100                 <?x ml version="1.0"?>

Number of entries: 55
Score distribution:
    5 +++++++++++++++++++ 19
    8 +++++ 5
   10 +++++++++++++++++ 17
   13 + 1
   15 ++++++++ 8
   18 + 1
   20 ++++ 4
Rule distribution:
  45x933100 23x942370 22x941100 5x942110 4x942440 3x933151 2x942330 2x932105 2x932150 2x942430 
Tests:
   53 success
    2 FAIL
```

**Comment by dune73** [comment]
Wow. This is impressive stuff and a detailed coverage of the problem. 

The closing tag is really cumbersome and if it complicates the regex then maybe we should leave it away - optionally trying to craft a future rule targeting the closing tag at PL3 or so.

Just my 2 cents.

**Comment by lifeforms** [comment]
Another thought is that my closing tag matching regexp might be vulnerable to regexp DoS because I am tracking necessarily from the start of the payload (because I need to assert that there's no `<?xml` there) - you might overload it by sending a megabyte of crap first.  I wish I knew more about regexp because there might be a better solution...

I think moving the closing tag to a different rule and higher PL would be appropriate. The closing tag conveys less of a threat than an opening tag (which you would need in most common injection cases) and is more problematic to match. Also it would lower FP to not check the closing tag, since `?>` can still arise often in binary data. But I don't know if this change might be too big for 3.0.1.

More opinions remain welcome.

**Comment by dune73** [comment]
I also thought about ReDos. I do not think you can avoid the megabyte of crap if you try to look at the start and the end of a variable in the same regex.

Creating a new rule at PL3 might introduce new FPs. I think we should postpone this to 3.1 and concentrate on handling the FPs of 933100 here.

**Comment by lifeforms** [comment]
Well I might be up for temporarily removing `?>` from rule 933100 in 3.0.1, and reintroduce a specific rule for it (at PL2) in 3.1. Such a new rule would have a new ruleID and could annoy people who have excluded 933100 now, that is true, so it would conflict with our release rules if we do this in 3.0.1. (PS: this is another reason we should introduce specific rule name tags and recommend people to exclude rule names instead of ruleIDs). Or we could defer this whole issue to 3.1. The pm to regexp conversion might already be a bit 'much' for a maintenance release.

I thought of another reason why detecting the closing tag is not so powerful. First, it's legal and actually recommended to omit the closing tag. This is a valid script:

```
<?php
echo 123;
```

If the attacker requires some garbage to be *after* the payload (like they are injecting in the middle of some JSON, or some other file format that is validated by the vulnerable application), they could start a comment:

```
<?php
echo 123; // garbagegarbagegarbage
```

Or if the garbage contains newlines they could call `__halt_compiler()` to abort parsing:

```
<?php
echo 123;
__halt_compiler();
garbagegarbagegarbage
```

Although I am detecting `__halt_compiler()` in rule 933150, so that trick would go wrong.

That said it's still somewhat useful to detect the closing tag.

**Comment by dune73** [comment]
I'm all for PL3 as I expect a lot of FPs with the closing tag and out of global rule design perspective. PL3 does not add much over PL2, while PL4 is a big step head. This became apparent in the tests before 3.0 was released. My research with ZHAW confirms this bias in the rules. Moving rules to PL3 takes away FPs from PL2 and gives a reason to run PL3.

I was not aware that you could omit the closing tag. That makes it even more important to separate the two.

**Comment by emphazer** [comment]
@lifeforms 
nice regex.
i love regex :-)

what do you think about this one instead.
i want to exclude xml only if no ; occurs until end of line.
````
yours:
@rx (?:<\?(?!xml\s)|<\?php|^(?!<\?xml\s).*\?>|\[(?:/|\\\\)?php\])

mine:
@rx (?:<\?(?!xml(?:[^;]+$)\s)|<\?php|^(?!<\?xml\s).*\?>|\[(?:/|\\\\)?php\])


````

but maybe this will lead to a problem with 
escaping of characters for different tags and attributes.

````
For tags:

 < &lt;
 > &gt; (only for compatibility)
 & &amp;


For attributes:

" &quot;
' &apos;

http://www.w3.org/TR/2008/REC-xml-20081126/#syntax

````

tried something like 
````
(?!&(lt|gt|amp|quot|apos);)
````

but no success so far

**Comment by lifeforms** [comment]
Discussion in IRC meeting:

- We temporarily drop the `?>` closing-tag detection from 3.0.1 since we've found no way to relieve this FP.
- We will reintroduce `?>` in a separate rule at higher paranoia level in 3.1, since we can't introduce new ruleIDs in 3.0.1 due to our release policy of not creating new FPs.

**Comment by dune73** [comment]
@lifeforms, are you working on this issue? Time is running out...

**Comment by lifeforms** [comment]
A PR is in #766. Please discuss there.

**Comment by lifeforms** [comment]
@emphazer I like your idea, but we need to work on it more, and time is very short, so I'd like to defer it. You point out a very useful improvement. It's true that if we open up the `<?xml` string, theoretically it's possible that an attacker could construct some interesting undetected attack, *if* the application contains a function or class named `xml`.

For instance, we now have the following false negatives:

```
<?xml ();         # the attacker can call function xml()
<?xml ::foo();    # the attacker can call static method xml::foo()
```

However for this to work, the vulnerable application should contain a target named `xml` which does something interesting, and the attacker needs to avoid a lot of PHP syntax. I figured this would be rare enough to warrant the inclusion. But maybe we can do more!

I think it's potentially clever to prevent this by looking for the `;`. I tried working with detecting a `;`. I think we don't have to watch out for encodings because the PHP engine does not decode them. But I'm having trouble with the basic idea. First, using a `;` is not necessary for the attacker. They can end the line with a comment and omit the `;`. Unfortunately these would be false negatives:

```
<?xml () // version="1.0"?>
<?xml ::test() // version="1.0"?>
```

On the other hand, an XML document with a `;` could now produce a false positive again:

```
<?xml version="1.0"?><foo>blah; blah; blah;</foo>
```

So it's hard. I think fiddling with the `;` turns out to be a bit too tricky for this late hour. If we can do something more to restrict the regexp later I would be all ears... Let's close this issue for housekeeping but we can restart the discussion in a different issue for the 3.1 version! By the way, the regexp I ultimately tried before giving up was:

```
(?:<\?(?!xml\s(?!.*;))|<\?php|\[(?:/|\\\\)?php\])
```
